test/gtest/core/test_allgather.cc

```diff
@@ -0,0 +1,149 @@
+            ucc_coll_args_t* coll = ctx->args;
+            if (coll->src.info.buffer) { /* no inplace */
+                UCC_CHECK(ucc_mc_free(coll->src.info.buffer, mem_type));
+            }
+            UCC_CHECK(ucc_mc_free(coll->dst.info.buffer, mem_type));
+            UCC_CHECK(ucc_mc_free(ctx->init_buf, UCC_MEMORY_TYPE_HOST));
+            free(coll);
+            free(ctx);
+        }
+        ctxs.clear();
+    }
+    void data_validate(UccCollCtxVec ctxs)
+    {
+        std::vector<uint8_t *> dsts(ctxs.size());
+
+        if (UCC_MEMORY_TYPE_HOST != mem_type) {
+            for (int r = 0; r < ctxs.size(); r++) {
+                UCC_CHECK(ucc_mc_alloc((void**)&dsts[r], ctxs[r]->rbuf_size,
+                                       UCC_MEMORY_TYPE_HOST));
+                ucc_mc_memcpy(dsts[r], ctxs[r]->args->dst.info.buffer,
+                              ctxs[r]->rbuf_size, UCC_MEMORY_TYPE_HOST, mem_type);
+            }
+        } else {
+            for (int r = 0; r < ctxs.size(); r++) {
+                dsts[r] = (uint8_t *)(ctxs[r]->args->dst.info.buffer);
+            }
+        }
+        for (int i = 0; i < ctxs.size(); i++) {
+            uint8_t *rbuf = dsts[i];
+            for (int r = 0; r < ctxs.size(); r++) {
+                size_t rank_size = ucc_dt_size((ctxs[r])->args->src.info.datatype) *
+                        (ctxs[r])->args->src.info.count;
+                for (int i = 0; i < rank_size; i++) {
+                    EXPECT_EQ(r, rbuf[r*rank_size + i]);
+                }
+            }
+        }
+        if (UCC_MEMORY_TYPE_HOST != mem_type) {
+            for (int r = 0; r < ctxs.size(); r++) {
+                ucc_mc_free((void*)dsts[r], UCC_MEMORY_TYPE_HOST);
+            }
+        }
+    }
+};
+
+class test_allgather_0 : public test_allgather,
+        public ::testing::WithParamInterface<Param_0> {};
+
+UCC_TEST_P(test_allgather_0, single_host)
+{
+    const int                 team_id  = std::get<0>(GetParam());
+    const ucc_datatype_t      dtype    = (ucc_datatype_t)std::get<1>(GetParam());
+    const ucc_memory_type_t   mem_type = std::get<2>(GetParam());
+    const int                 count    = std::get<3>(GetParam());
+    const gtest_ucc_inplace_t inplace  = std::get<4>(GetParam());
+    UccTeam_h                 team     = UccJob::getStaticTeams()[team_id];
+    int                       size     = team->procs.size();
+
+    set_inplace(inplace);
+    set_mem_type(mem_type);
+
+    UccCollCtxVec args = data_init(size, dtype, count);
+    UccReq    req(team, args);
+    req.start();
+    req.wait();
+    data_validate(args);
+    data_fini(args);
+}
+
+INSTANTIATE_TEST_CASE_P(
+    ,
+    test_allgather_0,
+    ::testing::Combine(
+        ::testing::Range(1, UccJob::nStaticTeams), // team_ids
+        ::testing::Range((int)UCC_DT_INT8, (int)UCC_DT_FLOAT64 + 1), // dtype
+#ifdef HAVE_CUDA
+        ::testing::Values(UCC_MEMORY_TYPE_HOST, UCC_MEMORY_TYPE_CUDA), // mem type
```

Comment:
If UCC is compiled with CUDA but the cuda is not available in runtime this will cause test failures. Need to skip CUDA tests if mc[cuda] is not available

Reply:
fixed with [set_mem_type](https://github.com/openucx/ucc/pull/114/files#diff-4ba093c65b5e5d06abb2606eb0817bd515a227e05bd1528b3bdc24d45a0e8e65R402-R409) function